autopush/tests/test_websocket.py

```diff
@@ -571,15 +571,118 @@ def fake_msg(data):
+                reactor.callLater(0.2, check_rotation, 0.2 + time_spent)
+                return
+
+            eq_(self.proto.ps.rotate_message_table, False)
+            d.callback(True)
+
         def check_result(msg):
+            # it's fine you've not connected in a while, but
+            # you should recycle your endpoints since they're probably
+            # invalid by now anyway.
+            eq_(msg["status"], 200)
+            eq_(msg["uaid"], orig_uaid)
+
+            # Wait to see that the message table gets rotated
+            reactor.callLater(0.2, check_rotation, 0.2)
+
+        self._check_response(check_result)
+        return d
+
+    def test_hello_tomorrow_provision_error(self):
+        orig_uaid = "deadbeef00000000abad1dea00000000"
+        router = self.proto.ap_settings.router
+        router.register_user(dict(
+            uaid=orig_uaid,
+            connected_at=ms_time(),
+            current_month="message_2016_3",
+            router_type="simplepush",
+        ))
+
+        # router.register_user returns (registered, previous
+        target_day = datetime.date(2016, 2, 29)
+        msg_day = datetime.date(2016, 3, 1)
+        msg_date = "{}_{}_{}".format(
+            self.proto.ap_settings._message_prefix,
+            msg_day.year,
+            msg_day.month)
+        msg_data = {
+            "router_type": "webpush",
+            "node_id": "http://localhost",
+            "last_connect": int(msg_day.strftime("%s")),
+            "current_month": msg_date,
+        }
+
+        def fake_msg(data):
+            return (True, msg_data, data)
+
+        mock_msg = Mock(wraps=db.Message)
+        mock_msg.fetch_messages.return_value = []
+        mock_msg.all_channels.return_value = (None, [])
+        self.proto.ap_settings.router.register_user = fake_msg
+        # massage message_tables to include our fake range
+        mt = self.proto.ps.settings.message_tables
+        for k in mt.keys():
+            del(mt[k])
+        mt['message_2016_1'] = mock_msg
+        mt['message_2016_2'] = mock_msg
+        mt['message_2016_3'] = mock_msg
+
+        patch_range = patch("autopush.websocket.randrange")
+        mock_patch = patch_range.start()
+        mock_patch.return_value = 1
+
+        def raise_error(*args):
+            raise ProvisionedThroughputExceededException(None, None)
+
+        self.proto.ap_settings.router.update_message_month = MockAssist([
+            raise_error,
+            Mock(),
+        ])
+
+        with patch.object(datetime, 'date',
+                          Mock(wraps=datetime.date)) as patched:
+            patched.today.return_value = target_day
+            self._connect()
+            self._send_message(dict(messageType="hello",
+                               uaid=orig_uaid,
+                               channelIDs=[],
+                               use_webpush=True))
+
+        d = Deferred()
+
+        def check_rotation(time_spent):
```

Comment:
you could reuse the other test's check_rotation w/ patch_range.stop on the deferred e.g. d.addBoth(lambda _: patch_range.stop())